tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -12426,11 +12426,11 @@ bool Sema::DiagnoseHLSLDecl(Declarator &D, DeclContext *DC, Expr *BitWidth,
          "otherwise this is called without checking language first");
 
   // NOTE: some tests may declare templates.
-  if (DC->isDependentContext()) return true;
+  if (DC->isDependentContext() && !getLangOpts().EnableTemplates) return true;
 
   DeclSpec::SCS storage = D.getDeclSpec().getStorageClassSpec();
   assert(!DC->isClosure() && "otherwise parser accepted closure syntax instead of failing with a syntax error");
-  assert(!DC->isDependentContext() && "otherwise parser accepted a template instead of failing with a syntax error");
+  assert((!DC->isDependentContext() || getLangOpts().EnableTemplates) && "otherwise parser accepted a template instead of failing with a syntax error");
```

Comment:
We should just delete this assert. It is impossible to hit because of the early return, and it is meaningless.